File: thruster.py
# Patrick McCorkell
# March 2021
# US Naval Academy
# Robotics and Control TSD
#

class Thruster:

	def setEvent(self):
		self._lock=1
		self.set_pw(self._base_pw)
	def clearEvent(self):
		self._lock=0

#	def cal_period(self,f):
#		self._period = 1000/f

#	def clampESC(self,n, minn, maxn):
	def clampESC(self,n):
		minn = self._max * -1
		maxn = self._max
		return min(max(n, minn), maxn)

	def set_pw(self,val):
		if (val>self._period):
########################### Add log functionality here:
			print("max period exceeded: " + str(val))
		else:
			newduty = round((val / self._period) * 4095)
			self._pca.duty(self._channel,newduty)
	def get_pw(self):
		g_pw = self._pca.duty(self._channel)/4095 * self._period
		return g_pw

	def set_speed(self,v):
		if (self._lock == 1):
			self.set_pw(self._base_pw)
		#elif (abs(v) > self._max):
########################### Add log functionality here:
		#	print("max speed exceeded: "+str(v))
		else:
			v*self._max
			vel = self.clampESC(v)
			# transform us to ms, and add to base 1.5 ms
			target_pw = (self._dir * vel / 1000) + self._base_pw
			# current_pw = self.get_pw()
			# diff_pw = abs(target_pw-current_pw)
			# if (diff_pw > self._tolerance_pw):
			self.set_pw(target_pw)
		return self.get_speed()
	def get_speed(self):
		sp = ((self.get_pw() - self._base_pw) * 1000) / self._max
		return sp

	def update_period(self):
		self._period = self._pca._period

	def __init__(self,pca,channel,direction):
		self._pca = pca						# pca9685 class
		self._channel = channel				# [0, 15] servo channel
		self._dir = direction				# [-1, 1] blade orientation
		self._lock = 0						# [0, 1] lock the thruster in safe 1.5ms
		self._max = self._pca._max			# us
		self._base_pw = 1.5					# ms
		self._tolerance_pw=0.0006			# ms
		self._period = self._pca._period	# ms
											# observed at 400Hz, +/- 1 bit in [0,4095] resolution results in precisely 0.0006000006000006497 ms difference in calculated pulsewidth

# For test purposes
# if __name__ == "thruster": 
# 	from pca9685 import PCA9685
# 	import busio
# 	from board import SCL,SDA

# 	i2c = busio.I2C(SCL,SDA)
# 	pca = PCA9685(i2c)

# 	pca.freq(400)

# 	servo_chan = 0
# 	direction = 1

# 	thr = Thruster(pca,servo_chan,direction)
# 	print("loaded thruster.py as main")

File: AY2019 codebase derived from Babelfish/00 Surface/thruster.cpp
/*
 * Thrust class for LPC1768
 * US Naval Academy
 * Robotics and Control TSD
 * Patrick McCorkell
 *
 * Created: 2020 Nov 23
 * 
 */

#include "mbed.h"
#include "thruster.h"

//Instantiation accepts PWM pin and direction of prop blade
//Direction is -1 or 1.
//1 for normal, -1 if blade reversed.
Thruster::Thruster(PinName pin, float dir) : _pwm(pin), _d(dir) {
    _lock=0;            // emergency lockout, default is 0
    _pin=pin;           // PWM pin
    _base_pw=1.5;       // 1.5ms
    set_pw(_base_pw);   // set PWM to 1.5ms
    _period=2.5;        // 2.5ms
    set_period(_period); // set period to 2.5ms (400Hz)
    _max=150;           // max PWM value in us
    //printf("Thruster: %f\r\n",d);
}

//Sets Event for Emergency Stop and sets lockout to set_speed() function.
void Thruster::setEvent() {
    _lock=1;            // set _lock flag to lockout set_speed functionality.
    set_pw(_base_pw);   // write the neutral PWM value.
}

//Clears Event for Emergency Stop of thruster and removes lockout from set_speed() function.
void Thruster::clearEvent() {
    _lock=0;            // set _lock flag back to 0, enabling set_speed functionality.
}

//Set PWM period in ms.
void Thruster::set_period(double thruster_time) {
    _period=thruster_time;
    _pwm.period(_period/1000);
}

//Set PWM pulsewidth in ms
void Thruster::set_pw(double thruster_pw) {
    double s_pw=(thruster_pw/1000);
    printf("log: set_pw: %f\r\n",s_pw);
    _pwm.pulsewidth(s_pw);
}

//Returns PWM pulsewidth in ms.
double Thruster::get_pw() {
    //read duty cycle times period
    double g_pw = (_pwm.read()*_period);
    //printf(" get_pw: %f, ",g_pw);
    return g_pw;
}

//Returns PWM output relative to 1.5ms.
double Thruster::get_speed() {
    double g_speed = (get_pw()-_base_pw);
    //printf("get_speed: %f, ",g_speed);
    return g_speed;
}

//formats PWM as an 2 uint16_t joined to make uint32_t for serial data streaming
//MSB uint16_t indicates direction, 0 for positive, 1 for negative.
//LSB uint16_t is 10us resolution of PWM
uint32_t Thruster::thruster_data() {
    double speed=get_speed();
    uint32_t dir=0x0;
    uint32_t data=0x0;
    if (speed<0) dir =0x00010000;
    data=static_cast<unsigned int>(abs(int(speed*100000)));
    data=data+dir;
    return data;
}

//Accepts adjustment to max value of pw [x,500] us for set_speed() function.
//Returns 1 if successful.
void Thruster::set_max(int new_max) {
    if (new_max<=500) {
        _max=new_max;
    }
}

int Thruster::get_max() {
    int buffer = _max;
    return buffer;
}

//Accepts adjustment to pw [-500,500] us that is added to 1.5ms.
//Returns 1 if successful.
int Thruster::set_speed(double speed_pw) {
    int returnval = 0;
    if (_lock==1) {
        set_pw(_base_pw);
    }
    else if (abs(speed_pw) > _max) {
        //print("max speed exceeded");
        returnval=0;
    }
    else {
        double tolerance_pw=0.001;
        double target_pw=(_d*speed_pw)+_base_pw;
        double current_pw=get_pw();
        double diff_pw=abs(target_pw-current_pw);
        if (diff_pw>tolerance_pw) {
            set_pw(target_pw);
            returnval = 1;
        }
    }
    return returnval;
}

